### ISSUE TITLE
Support @oneOf directive

### ISSUE THREAD

**Issue by bvanneerven** [async-graphql/async-graphql#766]
## Introduction

Hi! First of all thank you for this amazing library! It really is a joy to use! ✨🔥 

I have been following the discussion related to [Oneof Input Objects and Oneof Fields](https://github.com/graphql/graphql-spec/pull/825) in the GraphQL spec repository for a while now, and it seems that the community has settled on the `@oneOf` directive as solution. While it is not yet official, developers are already implementing this functionality, for example in an [Envelop plugin (JS)](https://www.envelop.dev/plugins/use-extended-validation#schema-directive-flow).

Since async-graphql at this point only supports `FIELD` directives, it seems that this functionality could not be implemented using the current API of async-graphql, which is why I opened this feature request.

## Requested feature

If async-graphql could implement an API with which we could create the following schema directive, then users of async-graphql could implement this functionality themselves until the GraphQL spec officialy releases it. Or, since it is Rust, async-graphql could already implement this in the library and hide it behind some feature flag.

```gql
directive @oneOf on INPUT_OBJECT | FIELD_DEFINITION
```

**Comment by sunli829** [comment]
I'm sorry to reply so late. `Oneof` looks great, I will implement it after the work at hand is completed. 🙂

**Comment by acelot** [comment]
@sunli829 any news on this?

**Comment by sunli829** [comment]
I have good news for you guys, I've started doing this, first of all will implement the oneOf input object.

**Comment by Erik1000** [comment]
Great to see you're working on this! Can you give us an ETA when `@oneOf` is available for output objects @sunli829?

**Comment by sunli829** [comment]
I've been thinking about how the `oneOf` field is defined in rust for a while.

```rust
struct Query;

#[Object]
impl Query {
    #[graphql(oneof)]
    async fn field(&self, a: Option<i32>, b: Option<i32>, c: Option<i32>) -> i32 {
        if let Some(a) = a {
            a
        } else if let Some(b) = b {
            b
        } else if let Some(c) = c {
            c
        } else {
              // This is strange, the function signature seems to allow `a`, `b`, `c` three parameters are all `None`.
              unreachable!()
        }
    }
}
```

So I think the following way would be better.

```rust
#[derive(OneOf)]
enum FieldArg {
    A(i32),
    B(i32),
    C(i32),
}

struct Query;

#[Object]
impl Query {
    #[graphql(oneof)]
    async fn field(&self, arg: FieldArg) -> i32 {
        match arg {
            A(a) => a,
            B(a) => a,
            C(a) => a,
        }
    }
}
```

If the `oneof` property is defined on the object field, it is allowed to have only one parameter, and the parameter type must be `OneOfObject`.

The generated schema will take all fields of `FieldArg` as arguments to `Query.field`.

```graphql
type Query {
    field(a: Int, b: Int, c: Int): Int! @oneof
}
```

**Comment by Erik1000** [comment]
I agree, the second approach is way better for type safety and easier to understand.
It's probably also more intuitive for output types since they could be written the same way.

**Comment by sunli829** [comment]
Released in `v3.0.33` 🙂

**Comment by Erik1000** [comment]
Just tested it and works like a charm. Tho I guess the [example in the docs](https://docs.rs/async-graphql/latest/async_graphql/derive.OneofObject.html#examples) is missing the `#[graphql(oneof)` property. Without that, the `oneof` directive is not set. Also, at least the documentation for envelop says that you should include 
```graphql
directive @oneOf on INPUT_OBJECT | FIELD_DEFINITION
```
in the schema which it's not.
